Update docs with new usage tracking info (#3626)
Parsl/parsl: docs/userguide/usage_tracking.rst

.. _label-usage-tracking:

Usage Statistics Collection
===========================

Parsl uses an **Opt-in** model for usage tracking, allowing users to decide if they wish to participate. Usage statistics are crucial for improving software reliability and help focus development and maintenance efforts on the most used components of Parsl. The collected data is used solely for enhancements and reporting and is not shared in its raw form outside of the Parsl team.

Why are we doing this?
----------------------

> The Parsl development team relies on funding from government agencies. To sustain this funding and advocate for continued support, it is essential to show that the scientific community benefits from these investments.

Reviewer: ```suggestion
The Parsl development team relies on funding from government agencies. To sustain this funding and advocate for continued support, it is essential to show that the research community benefits from these investments.
```

By opting in to share usage data, you actively support the ongoing development and maintenance of Parsl. (See:ref:`What is sent? <what-is-sent>` below).

Opt-In Model
------------

We use an **opt-in model** for usage tracking to respect user privacy and provide full control over shared information. We hope that developers and researchers will choose to send us this information. The reason is that we need this data - it is a requirement for funding.

Choose the data you share with Usage Tracking Levels.

**Usage Tracking Levels:**

> * **Level 1:** Basic information such as Python version, Parsl version, and platform details.

Reviewer: We might also want to say what "platform details" means

> * **Level 2:** Configuration information including provider, executor, and launcher names.

Reviewer: ```suggestion
* **Level 2:** Level 1 information and configuration information including provider, executor, and launcher names.
```

> * **Level 3:** Workflow execution details, including the number of applications run, failures, and execution time.

Reviewer: ```suggestion
* **Level 3:** Level 2 information and workflow execution details, including the number of applications run, failures, and execution time.
```

By enabling usage tracking, you support Parsl's development. 

> **To opt-in, set ``usage_tracking`` to the desired level (1, 2, or 3) in the configuration object (`parsl.config.Config`).**

Reviewer 2: the backquote formatting doesn't work when i build this locally, inside the ** bold block. you could remove the bold block.

Example:

.. code-block:: python3

    config = Config(
        executors=[
            HighThroughputExecutor(
                ...
            )
        ],
        usage_tracking=3
    )

.. _what-is-sent:

What is sent?
-------------

The data collected depends on the tracking level selected:

> * **Level 1:** Basic information such as Python version, Parsl version, and platform details.

Reviewer: ```suggestion
* **Level 1:** Only basic information such as Python version, Parsl version, and platform details.
```

Reviewer: We might want to say what "platform details" includes

> * **Level 2:** Configuration information including provider, executor, and launcher names.

Reviewer: ```suggestion
* **Level 2:** Level 1 information and configuration information including provider, executor, and launcher names.
```

> * **Level 3:** Workflow execution details, including the number of applications run, failures, and execution time.

Reviewer: ```suggestion
* **Level 3:** Level 2 information and workflow execution details, including the number of applications run, failures, and execution time.
```

    {
       "correlator":"6bc7484e-5693-48b2-b6c0-5889a73f7f4e",
       "execution_time":31,
       "components":[
          {
             "c":"parsl.dataflow.dflow.DataFlowKernel",
             "app_count":3,
             "app_fails":0
          },
          {
             "c":"parsl.config.Config",
             "executors_len":1,
             "dependency_resolver":false
          },
          "parsl.executors.threads.ThreadPoolExecutor"
       ],
       "end":1727156156
    }

**All messages sent are logged in the ``parsl.log`` file, ensuring complete transparency.**

How is the data sent?
---------------------

Data is sent using **UDP** to minimize the impact on workflow performance. While this may result in some data loss, it significantly reduces the chances of usage tracking affecting the software's operation.

The data is processed through AWS CloudWatch to generate a monitoring dashboard, providing valuable insights into usage patterns.

When is the data sent?
----------------------

Data is sent twice per run:

1. At the start of the script.
2. Upon script completion (for Tracking Level 3).

What will the data be used for?
-------------------------------

> The data will help answer key questions, including:

Reviewer: ```suggestion
The data will help the Parsl team understand Parsl usage and make development and maintenance decisions, including:
```

**Usage Tracking Levels:**

> * **Level 1:** Only basic information such as Python version, Parsl version, and platform details (Linux, MacOS etc.).

Reviewer: Is platform details just OS? If so, say so...  This is still too vague to me

Author: Yes. It's just the OS name. I'll make the change

**Usage Tracking Levels:**

> * **Level 1:** Only basic information such as Python version, Parsl version, and platform name (Linux, MacOS etc.).

Reviewer: ```suggestion
* **Level 1:** Only basic information such as Python version, Parsl version, and platform name (Linux, MacOS, etc.)
```

The data collected depends on the tracking level selected:

> * **Level 1:** Only basic information such as Python version, Parsl version, and platform name (Linux, MacOS etc.).

Reviewer: ```suggestion
* **Level 1:** Only basic information such as Python version, Parsl version, and platform name (Linux, MacOS, etc.)
```
